pymeasure/instruments/activetechnologies/AWG401x.py

```diff
@@ -0,0 +1,1145 @@
+            "SEQuence:ELEM<ent>:VOLTage:LOW<ch>?",
+            "SEQuence:ELEM<ent>:VOLTage:LOW<ch> %s",
+            """This property sets or returns the low voltage level of the
+            waveform.""",
+            validator=strict_range,
+            dynamic=True
+        )
+
+        voltage_low_max = Instrument.measurement(
+            "SEQuence:ELEM<ent>:VOLTage:LOW<ch>? MAXimum",
+            """This property queries the maximum low voltage level of the
+            waveform that can be set to the output waveform."""
+        )
+
+        voltage_low_min = Instrument.measurement(
+            "SEQuence:ELEM<ent>:VOLTage:LOW<ch>? MINimum",
+            """This property queries the minimum low voltage level of the
+            waveform that can be set to the output waveform."""
+        )
+
+        waveform = Instrument.control(
+            "SEQuence:ELEM<ent>:WAVeform<ch>?",
+            "SEQuence:ELEM<ent>:WAVeform<ch> %s",
+            """This property sets or returns the waveform. It’s possible select
+            a waveform only from those in the waveform list. In waveform list
+            are already present 10 predefined waveform: Sine, Ramp, Square,
+            Sync, DC, Gaussian, Lorentz, Haversine, Exp_Rise and Exp_Decay but
+            user can import in the list others customized waveforms.""",
+            validator=strict_discrete_set,
+            set_process=lambda v: f"\"{v}\"",
+            dynamic=True
+        )
+
+        def __init__(self, instrument, channel_number, sequence_number):
+            self.instrument = instrument
+            self.num_ch = channel_number
+            self.seq_num = sequence_number
+
+            self._special_names = self._setup_special_names()
+
+            self.waveform_values = list(self.instrument.waveforms.keys())
+            self.calculate_voltage_range()
+
+        def values(self, command, **kwargs):
+            command = command.replace("<ent>", str(self.seq_num))
+            command = command.replace("<ch>", str(self.num_ch))
+            return self.instrument.values(command, **kwargs)
+
+        def write(self, command):
+            command = command.replace("<ent>", str(self.seq_num))
+            command = command.replace("<ch>", str(self.num_ch))
+            self.instrument.write(command)
+
+        def calculate_voltage_range(self):
+            self.voltage_amplitude_values = list(map(
+                float,
+                [self.voltage_amplitude_max, self.voltage_amplitude_min]))
+
+            self.voltage_offset_values = list(map(
+                float,
+                [self.voltage_offset_max, self.voltage_offset_min]))
+
+            self.voltage_high_values = list(map(
+                float,
+                [self.voltage_high_max, self.voltage_high_min]))
+
+            self.voltage_low_values = list(map(
+                float,
+                [self.voltage_low_max, self.voltage_low_min]))
+
+
+class IEEE488DataBlock():
```

Comment:
I wonder if this class can be removed by using utilities provided by `pyvisa`.
See pymeasure/adapters/serial.py for some example

Reply:
@msmttchr For some context, this instrument can also accept data in the form of a text string of numbers separated by newlines, not just arrays of numbers. When I wrote a much more hacky implementation of this instrument (under the US distributer name Berkeley Nucleonics 675)  [here](https://github.com/neal-kepler/pymeasure/blob/master/pymeasure/instruments/berkeleynucleonics/bn675.py), I made the same choice. In light of your comment I think [write_binary](https://pyvisa.readthedocs.io/en/latest/introduction/rvalues.html) with datatype='s' may actually work, I didn't realize at the time that you could pass raw strings. @garzetti can correct me if they have already tried.

Reply:
I have removed the `IEEE488DataBlock` class, as suggested by @msmttchr , and used `from_ieee_block` and `to_ieee_block` methods from pyvisa.util . 
As suggested by @physicsninja , probably using `query_binary_values` and `write_binary_values` from pyvisa will work, but in my opinion, it is better to modify the `pymeasure.adapter.visa` class to include these two methods instead of calling it directly from pyvisa.

Reply:
What about passing the keyword parameter `datatype='s'` to `binary_values` and `write_binary_values` ? Does this the trick?

Reply:
The `write_binary_values` will work because it simply call the pyvisa `write_binary_values` method in which specifying `datatype='s'` works correctly. I can not use `binary_values` because it does not expose directly the pyvisa `query_binary_values` in which i must use `header_fmt='ieee'` and `datatype='d'` (yes, the data type are different read direction).
I can suggest two solution:
 * Use directly the `connection` object in visa adapter (i.e., `self.instrument.connection.query_binary_values(...)` )
 * Modify the visa adapter to expose directly a `query_binary_values` method like the `write_binary_values` one

Reply:
@physicsninja and @garzetti thanks for your valuable feedback and clarification.
Let me share my opinion below.

> The `write_binary_values` will work because it simply call the pyvisa `write_binary_values` method in which specifying `datatype='s'` works correctly. I can not use `binary_values` because it does not expose directly the pyvisa `query_binary_values` in which i must use `header_fmt='ieee'` and `datatype='d'` (yes, the data type are different read direction). I can suggest two solution:
> 
>     * Use directly the `connection` object in visa adapter (i.e., `self.instrument.connection.query_binary_values(...)` )
> 
This has the disadvantage to restrict support (when using GPIB interface) to adapters supported by pyvisa (e.g. cheap Prologix is not supported), but the instrument comes by default with LAN interface, so it should be OK.

>     * Modify the visa adapter to expose directly a `query_binary_values` method like the `write_binary_values` one
The adapter are under restructuring, so, I would avoid to add more methods, especially because this needs to be kept consistent with other adapters. I would be in favor of doing this extension in a dedicated PR. If you have time have a look at the ongoing discussion in #617.

In summary, I would vote for the first option you suggested.

Reply:
@msmttchr thanks for your suggestion. Modified to use the first solution.